### ISSUE TITLE
Unify _msge_with_gradient_*

### ISSUE THREAD

**Issue by cle1109** [scot-dev/scot#136]
We should refactor `_msge_with_gradient_underdetermined` and `_msge_with_gradient_overdetermined` in `var.py` into one function.

**Comment by mbillingr** [comment]
How do you propose to proceed?

Note that these functions are performance critical. Originally, I split them to avoid comparisons inside the optimization loop... but back then I was still thinking like a C/C++ programmer. Anyway, whatever we do to these functions, performance should not suffer.

**Comment by cle1109** [comment]
Definitely. The whole msge thing seems a bit convoluted. Do you think a comparison affects performance? How much can we merge the over/underdetermined functions? I guess the variable names have a special meaning, but to me bae, baee, baecc, j, k, l don't mean anything so we could use identical names for both versions. Then the only difference seems to be the order of operations (and transposition).

However, now I'm not sure if merging really makes the whole thing more readable.

**Comment by mbillingr** [comment]
The purpose of using different order of operations was to get smaller intermediate matrices, saving memory and arithmetic operations. To be honest, I'm not even sure if there is a significant performance difference between both functions. We should verify this with a benchmark before doing anything else. Then we can use this benchmark to check if the refactoring affected performance.

The variable names were chosen so that they can be read as mathematical formulas. `a, b, c, d, e` are matrices. Then, for example, variable `abc` would be the matrix product `A*B*C` (ignoring transposes).
